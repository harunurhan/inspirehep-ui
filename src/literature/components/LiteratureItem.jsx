import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { Map } from 'immutable';

import ResultItem from '../../common/components/ResultItem';
import AuthorList from './AuthorList';
import LiteratureDate from './LiteratureDate';
import ArxivEprintList from './ArxivEprintList';
import DOIList from './DOIList';
import CollapsableAbstract from './CollapsableAbstract';
import ReportNumberList from './ReportNumberList';
import PublicationInfoList from './PublicationInfoList';
import ArxivPdfDownloadAction from './ArxivPdfDownloadAction';
import CiteModalAction from './CiteModalAction';
import ListItemAction from '../../common/components/ListItemAction';
import Latex from '../../common/components/Latex';

class LiteratureItem extends Component {
  render() {
    const { metadata, display } = this.props;

    const title = metadata.getIn(['titles', 0, 'title']);
    const authors = metadata.get('authors');

    const arxivId = metadata.getIn(['arxiv_eprints', 0, 'value']);
    const recordId = metadata.get('control_number');
    const citationCount = metadata.get('citation_count');
    const referenceCount = display.get('number_of_references');

    const date = metadata.get('earliest_date');
    const publicationInfo = metadata.get('publication_info');
    const eprints = metadata.get('arxiv_eprints');
    const dois = metadata.get('dois');
    const reportNumbers = metadata.get('report_numbers');
    const abstract = metadata.getIn(['abstracts', 0, 'value']);

    return (
      <ResultItem
        title={<Latex>{title}</Latex>}
        description={(
          <div>
            {recordId && <AuthorList recordId={recordId} authors={authors} />}
            <LiteratureDate date={date} />
          </div>
        )}
        actions={[
          arxivId && <ArxivPdfDownloadAction arxivId={arxivId} />,
          recordId && <CiteModalAction recordId={recordId} />,
          recordId && citationCount && <ListItemAction iconType="logout" text={`${citationCount} citations`} href={`/literature/${recordId}#citations`} />,
          recordId && referenceCount && <ListItemAction iconType="login" text={`${referenceCount} references`} href={`/literature/${recordId}#citations`} />,
        ].filter(action => action != null)}
      >
        <PublicationInfoList publicationInfo={publicationInfo} />
        <ArxivEprintList eprints={eprints} />
        <DOIList dois={dois} />
        <ReportNumberList reportNumbers={reportNumbers} />
        <CollapsableAbstract abstract={abstract} collapsable />
      </ResultItem>
    );
  }
}

LiteratureItem.propTypes = {
  metadata: PropTypes.instanceOf(Map).isRequired,
  display: PropTypes.instanceOf(Map).isRequired,
};


export default LiteratureItem;
