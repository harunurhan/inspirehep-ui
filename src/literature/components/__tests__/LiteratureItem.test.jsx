import React from 'react';
import { shallow } from 'enzyme';
import { fromJS } from 'immutable';

import LiteratureItem from '../LiteratureItem';

describe('LiteratureItem', () => {
  it('renders with all props set, including sub props', () => {
    const metadata = fromJS({
      titles: [
        { title: 'test' },
      ],
      authors: [
        { full_name: 'Test, Author' },
      ],
      arxiv_eprints: [
        { value: '1234567890' },
      ],
      control_number: 12345,
      citation_count: 1,
      publicaion_info: [
        { journal_title: 'Test Jornal' },
      ],
      dois: [
        { value: '12345679.1234.123' },
      ],
      report_numbers: [
        { value: 'ABCD-AB-CD-1234-123' },
      ],
      abstract: [
        { value: 'Test Abstract' },
      ],
    });
    const display = fromJS({
      date: 'Jun 7, 1993',
      number_of_references: 2,
    });
    const wrapper = shallow((
      <LiteratureItem metadata={metadata} display={display} />
    ));
    expect(wrapper).toMatchSnapshot();
  });

  it('does not render citation and reference count if they are missing', () => {
    const metadata = fromJS({
      control_number: 12345,
    });
    const display = fromJS({});
    const wrapper = shallow((
      <LiteratureItem metadata={metadata} display={display} />
    ));
    expect(wrapper).toMatchSnapshot();
  });

  it('does not render cite, citation count and reference count if control_number is missing', () => {
    const metadata = fromJS({
      citation_count: 1,
    });
    const display = fromJS({
      number_of_references: 2,
    });
    const wrapper = shallow((
      <LiteratureItem metadata={metadata} display={display} />
    ));
    expect(wrapper).toMatchSnapshot();
  });

  it('does not arxiv pdf download action if there is no eprint value', () => {
    const metadata = fromJS({
      control_number: 12345,
    });
    const display = fromJS({});
    const wrapper = shallow((
      <LiteratureItem metadata={metadata} display={display} />
    ));
    expect(wrapper).toMatchSnapshot();
  });
});
