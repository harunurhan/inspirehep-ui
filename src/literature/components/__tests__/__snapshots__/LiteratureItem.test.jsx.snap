// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LiteratureItem does not arxiv pdf download action if there is no eprint value 1`] = `
<ResultItem
  actions={
    Array [
      <CiteModalAction
        recordId={12345}
    />,
    ]
  }
  description={
    <div>
      <AuthorList
        authors={
          Immutable.List [
          ]
        }
        limit={3}
        recordId={12345}
      />
      <LiteratureDate
        date={null}
      />
    </div>
  }
  title={<Latex />}
>
  <PublicationInfoList
    publicationInfo={null}
  />
  <ArxivEprintList
    eprints={null}
  />
  <DOIList
    dois={null}
  />
  <ReportNumberList
    reportNumbers={null}
  />
  <CollapsableAbstract
    abstract={null}
  />
</ResultItem>
`;

exports[`LiteratureItem does not render citation and reference count if they are missing 1`] = `
<ResultItem
  actions={
    Array [
      <CiteModalAction
        recordId={12345}
    />,
    ]
  }
  description={
    <div>
      <AuthorList
        authors={
          Immutable.List [
          ]
        }
        limit={3}
        recordId={12345}
      />
      <LiteratureDate
        date={null}
      />
    </div>
  }
  title={<Latex />}
>
  <PublicationInfoList
    publicationInfo={null}
  />
  <ArxivEprintList
    eprints={null}
  />
  <DOIList
    dois={null}
  />
  <ReportNumberList
    reportNumbers={null}
  />
  <CollapsableAbstract
    abstract={null}
  />
</ResultItem>
`;

exports[`LiteratureItem does not render cite, citation count and reference count if control_number is missing 1`] = `
<ResultItem
  actions={Array []}
  description={
    <div>
      <LiteratureDate
        date={null}
      />
    </div>
  }
  title={<Latex />}
>
  <PublicationInfoList
    publicationInfo={null}
  />
  <ArxivEprintList
    eprints={null}
  />
  <DOIList
    dois={null}
  />
  <ReportNumberList
    reportNumbers={null}
  />
  <CollapsableAbstract
    abstract={null}
  />
</ResultItem>
`;

exports[`LiteratureItem renders with all props set, including sub props 1`] = `
<ResultItem
  actions={
    Array [
      <ArxivPdfDownloadAction
        arxivId="1234567890"
    />,
      <CiteModalAction
        recordId={12345}
    />,
      <ListItemAction
        href="/literature/12345#citations"
        iconType="logout"
        onClick={undefined}
        target={undefined}
        text="1 citations"
    />,
      <ListItemAction
        href="/literature/12345#references"
        iconType="login"
        onClick={undefined}
        target={undefined}
        text="2 references"
    />,
    ]
  }
  description={
    <div>
      <AuthorList
        authors={
          Immutable.List [
            Immutable.Map {
              full_name: "Test, Author",
            },
          ]
        }
        limit={3}
        recordId={12345}
      />
      <LiteratureDate
        date="Jun 7, 1993"
      />
    </div>
  }
  title={
    <Latex>
      test
    </Latex>
  }
>
  <PublicationInfoList
    publicationInfo={null}
  />
  <ArxivEprintList
    eprints={
      Immutable.List [
        Immutable.Map {
          value: "1234567890",
        },
      ]
    }
  />
  <DOIList
    dois={
      Immutable.List [
        Immutable.Map {
          value: "12345679.1234.123",
        },
      ]
    }
  />
  <ReportNumberList
    reportNumbers={
      Immutable.List [
        Immutable.Map {
          value: "ABCD-AB-CD-1234-123",
        },
      ]
    }
  />
  <CollapsableAbstract
    abstract={null}
  />
</ResultItem>
`;
